File: packages/codex-subagent-kit/src/catalog-sync.ts
import { cpSync, existsSync, mkdirSync, readdirSync, rmSync, writeFileSync } from "node:fs";
import { basename, resolve } from "node:path";

import {
  DEFAULT_SYNC_SOURCE_NAME,
  resolveGlobalSourceCategoriesDir,
  resolveProjectSourceCategoriesDir,
} from "./paths";

const VOLTAGENT_TREE_API_URL =
  "https://api.github.com/repos/VoltAgent/awesome-codex-subagents/git/trees/main?recursive=1";
const VOLTAGENT_RAW_BASE_URL =
  "https://raw.githubusercontent.com/VoltAgent/awesome-codex-subagents/main/";

export class CatalogSyncError extends Error {}

export interface CatalogSyncResult {
  sourceName: string;
  targetRoot: string;
  sourceLabel: string;
  copiedPaths: string[];
}

function normalizeSourceName(sourceName: string): string {
  const candidate = sourceName.trim().toLowerCase();
  if (!candidate) {
    throw new CatalogSyncError("catalog sync requires a non-empty source name");
  }
  if (!/^[a-z0-9][a-z0-9_-]*$/.test(candidate)) {
    throw new CatalogSyncError(
      "source names may contain only lowercase letters, digits, hyphens, and underscores",
    );
  }
  return candidate;
}

function resolveCategoriesRoot(sourceRoot: string): string {
  const root = resolve(sourceRoot);
  if (existsSync(root) && basename(root) === "categories") {
    return root;
  }
  const nested = resolve(root, "categories");
  if (existsSync(nested)) {
    return nested;
  }
  throw new CatalogSyncError(`expected a categories/ directory but found: ${sourceRoot}`);
}

function prepareTargetRoot(targetRoot: string): void {
  mkdirSync(resolve(targetRoot, ".."), { recursive: true });
  if (existsSync(targetRoot)) {
    rmSync(targetRoot, { recursive: true, force: true });
  }
  mkdirSync(targetRoot, { recursive: true });
}

function listRelativeCatalogFiles(root: string): string[] {
  const discovered: string[] = [];

  function walk(current: string): void {
    for (const entry of readdirSync(current, { withFileTypes: true })) {
      const absolutePath = resolve(current, entry.name);
      if (entry.isDirectory()) {
        walk(absolutePath);
        continue;
      }
      if (!entry.isFile()) {
        continue;
      }
      if (entry.name === "README.md" || entry.name.endsWith(".toml")) {
        discovered.push(absolutePath.slice(root.length + 1));
      }
    }
  }

  walk(root);
  return discovered.sort();
}

function copyLocalSource(sourceRoot: string, targetRoot: string): string[] {
  const categoriesRoot = resolveCategoriesRoot(sourceRoot);
  const copiedPaths: string[] = [];
  for (const relativePath of listRelativeCatalogFiles(categoriesRoot)) {
    const sourcePath = resolve(categoriesRoot, relativePath);
    const destination = resolve(targetRoot, relativePath);
    mkdirSync(resolve(destination, ".."), { recursive: true });
    cpSync(sourcePath, destination);
    copiedPaths.push(destination);
  }
  return copiedPaths;
}

async function fetchJson(url: string): Promise<Record<string, unknown>> {
  const response = await fetch(url, {
    headers: {
      Accept: "application/vnd.github+json",
      "User-Agent": "codex-subagent-kit",
    },
  });
  if (!response.ok) {
    throw new CatalogSyncError(`failed to fetch upstream catalog metadata: ${response.status}`);
  }
  return (await response.json()) as Record<string, unknown>;
}

async function fetchText(url: string): Promise<string> {
  const response = await fetch(url, {
    headers: { "User-Agent": "codex-subagent-kit" },
  });
  if (!response.ok) {
    throw new CatalogSyncError(`failed to fetch upstream catalog file: ${url}`);
  }
  return response.text();
}

async function downloadVoltAgentSource(targetRoot: string): Promise<string[]> {
  const payload = await fetchJson(VOLTAGENT_TREE_API_URL);
  const tree = payload.tree;
  if (!Array.isArray(tree)) {
    throw new CatalogSyncError("unexpected upstream catalog metadata payload");
  }

  const upstreamPaths = tree
    .map((entry) => (entry && typeof entry === "object" ? entry.path : undefined))
    .filter((path): path is string => typeof path === "string" && path.startsWith("categories/"))
    .filter((path) => path.endsWith(".toml") || path.endsWith("README.md"))
    .sort();

  if (upstreamPaths.length === 0) {
    throw new CatalogSyncError("upstream catalog did not expose any categories files");
  }

  const copiedPaths: string[] = [];
  for (const upstreamPath of upstreamPaths) {
    const relativePath = upstreamPath.replace(/^categories\//, "");
    const destination = resolve(targetRoot, relativePath);
    mkdirSync(resolve(destination, ".."), { recursive: true });
    writeFileSync(destination, await fetchText(`${VOLTAGENT_RAW_BASE_URL}${upstreamPath}`), "utf8");
    copiedPaths.push(destination);
  }

  return copiedPaths;
}

export async function syncCatalog(options: {
  projectRoot: string;
  scope: "project" | "global";
  homeDir?: string;
  sourceName?: string;
  sourceRoot?: string;
}): Promise<CatalogSyncResult> {
  const sourceName = normalizeSourceName(options.sourceName ?? DEFAULT_SYNC_SOURCE_NAME);
  const targetRoot =
    options.scope === "project"
      ? resolveProjectSourceCategoriesDir(options.projectRoot, sourceName)
      : resolveGlobalSourceCategoriesDir(sourceName, options.homeDir);

  prepareTargetRoot(targetRoot);

  let copiedPaths: string[];
  let sourceLabel: string;
  if (options.sourceRoot) {
    copiedPaths = copyLocalSource(options.sourceRoot, targetRoot);
    sourceLabel = resolveCategoriesRoot(options.sourceRoot);
  } else {
    if (sourceName !== DEFAULT_SYNC_SOURCE_NAME) {
      throw new CatalogSyncError(
        "remote sync is only supported for the default voltagent source without --source-root",
      );
    }
    copiedPaths = await downloadVoltAgentSource(targetRoot);
    sourceLabel = "VoltAgent/awesome-codex-subagents@main";
  }

  if (copiedPaths.length === 0) {
    throw new CatalogSyncError("catalog sync did not copy any README or TOML files");
  }

  return {
    sourceName,
    targetRoot,
    sourceLabel,
    copiedPaths,
  };
}

File: src/codex_subagent_kit/catalog_sync.py
from __future__ import annotations

import json
import re
import shutil
from dataclasses import dataclass
from pathlib import Path
from urllib.error import URLError
from urllib.request import Request, urlopen

from .app_paths import (
    DEFAULT_SYNC_SOURCE_NAME,
    resolve_global_source_categories_dir,
    resolve_project_source_categories_dir,
)


VOLTAGENT_TREE_API_URL = (
    "https://api.github.com/repos/VoltAgent/awesome-codex-subagents/git/trees/main?recursive=1"
)
VOLTAGENT_RAW_BASE_URL = "https://raw.githubusercontent.com/VoltAgent/awesome-codex-subagents/main/"
SYNCABLE_SUFFIXES = (".toml", "README.md")


class CatalogSyncError(RuntimeError):
    """Raised when catalog sync cannot complete."""


@dataclass(frozen=True)
class CatalogSyncResult:
    source_name: str
    target_root: Path
    source_label: str
    copied_paths: list[Path]


def _normalize_source_name(source_name: str) -> str:
    candidate = source_name.strip().lower()
    if not candidate:
        raise CatalogSyncError("catalog sync requires a non-empty source name")
    if not re.fullmatch(r"[a-z0-9][a-z0-9_-]*", candidate):
        raise CatalogSyncError("source names may contain only lowercase letters, digits, hyphens, and underscores")
    return candidate


def _resolve_categories_root(source_root: Path) -> Path:
    root = source_root.resolve()
    if root.is_dir() and root.name == "categories":
        return root
    nested = root / "categories"
    if nested.is_dir():
        return nested
    raise CatalogSyncError(f"expected a categories/ directory but found: {source_root}")


def _prepare_target_root(target_root: Path) -> None:
    target_root.parent.mkdir(parents=True, exist_ok=True)
    if target_root.exists():
        shutil.rmtree(target_root)
    target_root.mkdir(parents=True, exist_ok=True)


def _iter_relative_catalog_files(root: Path) -> list[Path]:
    files: list[Path] = []
    for path in sorted(root.rglob("*")):
        if not path.is_file():
            continue
        if path.name == "README.md" or path.suffix == ".toml":
            files.append(path.relative_to(root))
    return files


def _copy_local_source(*, source_root: Path, target_root: Path) -> list[Path]:
    categories_root = _resolve_categories_root(source_root)
    copied_paths: list[Path] = []
    for relative_path in _iter_relative_catalog_files(categories_root):
        source_path = categories_root / relative_path
        destination = target_root / relative_path
        destination.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy2(source_path, destination)
        copied_paths.append(destination)
    return copied_paths


def _fetch_json(url: str) -> dict[str, object]:
    request = Request(
        url,
        headers={
            "Accept": "application/vnd.github+json",
            "User-Agent": "codex-subagent-kit",
        },
    )
    try:
        with urlopen(request) as response:
            return json.load(response)
    except URLError as exc:
        raise CatalogSyncError(f"failed to fetch upstream catalog metadata: {exc}") from exc


def _fetch_text(url: str) -> str:
    request = Request(url, headers={"User-Agent": "codex-subagent-kit"})
    try:
        with urlopen(request) as response:
            return response.read().decode("utf-8")
    except URLError as exc:
        raise CatalogSyncError(f"failed to fetch upstream catalog file: {url}") from exc


def _download_voltagent_source(*, target_root: Path) -> list[Path]:
    payload = _fetch_json(VOLTAGENT_TREE_API_URL)
    tree = payload.get("tree")
    if not isinstance(tree, list):
        raise CatalogSyncError("unexpected upstream catalog metadata payload")

    catalog_paths: list[str] = []
    for entry in tree:
        if not isinstance(entry, dict):
            continue
        path = entry.get("path")
        if not isinstance(path, str) or not path.startswith("categories/"):
            continue
        if path.endswith(".toml") or path.endswith("README.md"):
            catalog_paths.append(path)

    if not catalog_paths:
        raise CatalogSyncError("upstream catalog did not expose any categories files")

    copied_paths: list[Path] = []
    for upstream_path in sorted(catalog_paths):
        relative_path = Path(upstream_path).relative_to("categories")
        destination = target_root / relative_path
        destination.parent.mkdir(parents=True, exist_ok=True)
        destination.write_text(
            _fetch_text(f"{VOLTAGENT_RAW_BASE_URL}{upstream_path}"),
            encoding="utf-8",
        )
        copied_paths.append(destination)
    return copied_paths


def sync_catalog(
    *,
    project_root: Path,
    scope: str,
    source_name: str = DEFAULT_SYNC_SOURCE_NAME,
    source_root: Path | None = None,
) -> CatalogSyncResult:
    normalized_source_name = _normalize_source_name(source_name)

    if scope == "project":
        target_root = resolve_project_source_categories_dir(project_root, normalized_source_name)
    elif scope == "global":
        target_root = resolve_global_source_categories_dir(normalized_source_name)
    else:
        raise CatalogSyncError(f"unsupported scope: {scope}")

    _prepare_target_root(target_root)
    if source_root is not None:
        copied_paths = _copy_local_source(source_root=source_root, target_root=target_root)
        source_label = str(_resolve_categories_root(source_root))
    else:
        if normalized_source_name != DEFAULT_SYNC_SOURCE_NAME:
            raise CatalogSyncError(
                "remote sync is only supported for the default voltagent source without --source-root"
            )
        copied_paths = _download_voltagent_source(target_root=target_root)
        source_label = "VoltAgent/awesome-codex-subagents@main"

    if not copied_paths:
        raise CatalogSyncError("catalog sync did not copy any README or TOML files")

    return CatalogSyncResult(
        source_name=normalized_source_name,
        target_root=target_root,
        source_label=source_label,
        copied_paths=copied_paths,
    )
